# WES Variant Calling Pipeline

This repository contains a complete pipeline for variant calling from **Whole Exome Sequencing (WES)** data using tools like `FastQC`, `fastp`, `BWA`, `Picard` and `GATK`.
This is part of a project that involved variant calling from WES of families with primary open angle glaucoma.
Variant calling was done parallely on split bam files based on chromosomes and then merged together for speed and memory efficiency. If your aim does not allow that please stick to regular HaplotypeCaller or HaplotypeCallerSpark.
Change the arguments specified to allocate resources as per your needs. 

---

## ğŸ“ Project Structure

```
.
â”œâ”€â”€ data/                   # Raw sequencing data (not tracked)
â”œâ”€â”€ output/                 # Output directory (not tracked)
â”œâ”€â”€ tools/                 # Tools and binaries
â”œâ”€â”€ Ref/                   # Reference files (e.g., hg38, dbs)
â”œâ”€â”€ variant_calling.sh     # Main pipeline script
â”œâ”€â”€ data.txt               # Sample names (e.g., SRR32421545)
â”œâ”€â”€ .gitignore
â””â”€â”€ README.md
```



## âš™ï¸ Tools Used

- `FastQC` â€“ Quality check
- `fastp` â€“ Trimming and adapter removal
- `BWA-MEM` â€“ Read alignment
- `Picard` â€“ Mark duplicates and read group assignment
- `GATK` â€“ Variant calling and BQSR

---

## ğŸ” Pipeline Steps

1. **Input:** Paired-end FASTQ files (NovaSeq, Hybrid Selection) or binary SRA file
2. **Quality Control:** `FastQC`, `fastp`
3. **Alignment:** `bwa mem` â†’ `SAM` â†’ `BAM` â†’ sorted BAM
4. **Mark Duplicates:** `Picard`
5. **Base Quality Recalibration (BQSR):** `GATK BaseRecalibrator` and `ApplyBQSR`
6. **Variant Calling:** `GATK HaplotypeCaller` (parallel per chromosome)

---

## ğŸš€ How to Run

```bash
bash variant_calling.sh
Make sure to update data.txt with your SRA run accessions or sample names.

ğŸ“¦ Requirements
Linux with bash
Java â‰¥ 8
Python 3

Tools installed in tools/ directory and added to $PATH

ğŸ“ Notes
This pipeline is optimized for human WES data.

Large files like .fastq, .bam, and .vcf.gz are excluded via .gitignore.

For known sites in BQSR, download:

1000G_phase1.snps.high_confidence.hg38.vcf.gz

Mills_and_1000G_gold_standard.indels.hg38.vcf.gz

ğŸ“§ Contact
YadhuKrishnan
Email: yadhuryk@gmail.com
GitHub: @krishnanyadhu
